### PR TITLE
fix(completion): wrong select after leader change

### LINKED ISSUE #16128: after new refactor the select is wrong after leader add/delete
### Steps to reproduce

buffer with content `Foo Find five film` 
1:  `<C-x><C-P>`
2: intput F

Find not select 

this is a simple test case still have other cases i have find .  All the tests passed which led me to mistakenly believe that the refactoring was effective. In fact the test cases were lacking in many cases. I am trying to fix.

### Expected behaviour

work as before

### Version of Vim

build from source

### Environment

macos 

### Logs and stack traces

```shell
NAN
```

### PR DESCRIPTION
**Pull request by glepnir**
compl_shown_match update not correct afte refacot
this will make select wrong item after leader change we can generate compl_match_arraysie within before update compl_shown_match range, split generate compl_match_array into range match_head

Fix https://github.com/vim/vim/issues/16128

### CODE REVIEW

File: src/insexpand.c [hunk 1]
```diff
@@ -1258,8 +1269,76 @@ ins_compl_build_pum(void)
...
+		    shown_match_ok = TRUE;
+		}
+
+		// If there is no "no select" condition and the max fuzzy
+		// score is positive, or there is no completion leader or the
+		// leader length is zero, mark the shown match as valid and
+		// reset the current index.
+		if (!compl_no_select
+			&& (max_fuzzy_score > 0
+				|| (compl_leader.string == NULL || compl_leader.length == 0)))
```
Comment (chrisbra) [inline src/insexpand.c]:
```suggestion
				|| compl_leader.length == 0))
```
I don't think compl_leader.length can be positive while compl_leader.string is NULL.

File: src/insexpand.c [hunk 2]
```diff
@@ -1258,8 +1269,76 @@ ins_compl_build_pum(void)
...
+		if (!shown_match_ok && compl == compl_shown_match && !compl_no_select)
+		{
+		    cur = i;
+		    shown_match_ok = TRUE;
+		}
+
+		// If there is no "no select" condition and the max fuzzy
+		// score is positive, or there is no completion leader or the
+		// leader length is zero, mark the shown match as valid and
+		// reset the current index.
```
Comment (chrisbra) [inline src/insexpand.c]: Is that still correct? We are not resetting the index

File: src/insexpand.c (hunk 1 above)
Comment (glepnir) [reply src/insexpand.c]: leader can be an empty string the length is 0 but is not NULL

File: src/insexpand.c (hunk 2 above)
Comment (glepnir) [reply src/insexpand.c]: Nope we don't need this now and i will improve shown match when fuzzy in next pr.

**Comment by chrisbra** [comment]
ah I see, thanks

**Pull request closed without merging**